gossip/tx_scrambler_test.go

```diff
@@ -0,0 +1,342 @@
+package gossip
+
+import (
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"golang.org/x/exp/rand"
+	"math"
+	"math/big"
+	"reflect"
+	"testing"
+)
+
+func TestTxScrambler_AnalyseEntryList_RemovesDuplicateTransactions(t *testing.T) {
+	entries := createScramblerTestInputRepeatedHashes(2, 10)
+	entries, _, _ = analyseEntryList(entries)
+
+	checkDuplicateHashes(t, entries)
+}
+
+func TestTxScrambler_UnifyEntries_SaltCreationIsDeterministic(t *testing.T) {
+	entries := createScramblerTestInputRepeatedAddr(2, 10)
+	_, wantedSalt, _ := analyseEntryList(entries)
+	for _ = range 10 {
+		shuffleEntries(entries)
+		_, gotSalt, _ := analyseEntryList(entries)
+		if gotSalt != wantedSalt {
+			t.Fatal("incorrect salt - salt creation is not deterministic")
+		}
+	}
+
+}
+
+func TestTxScrambler_AnalyseEntryList_ReportsDuplicateAddresses(t *testing.T) {
+	entries := createScramblerTestInputRepeatedAddr(2, 10)
+	_, _, hasDuplicateAddresses := analyseEntryList(entries)
+	if !hasDuplicateAddresses {
+		t.Error("entries have duplicate addresses")
+	}
+
+	entries = createScramblerTestInputRepeatedHashes(2, 10)
+	_, _, hasDuplicateAddresses = analyseEntryList(entries)
+	if hasDuplicateAddresses {
+		t.Error("entries does not have duplicate addresses")
+	}
+}
+
+func TestTxScrambler_ScrambleTransactions_ScrambleIsDeterministic(t *testing.T) {
+	res1 := createScramblerTestInputRepeatedAddr(2, 10)
+	res2 := deepCopyEntries(res1)
+	// shuffle one array
+	shuffleEntries(res2)
+
+	for i := 0; i < 10; i++ {
+		salt := createRandomSalt()
+		scrambleTransactions(res1, salt)
+		for j := 0; j < 10; j++ {
+			scrambleTransactions(res2, salt)
```

Comment:
I think `res2` should be shuffled each time before calling `scrambleTransactions`